<a href="https://colab.research.google.com/github/syh0397/Statistics_python/blob/main/10_%EB%8B%A4%EC%A4%91%ED%9A%8C%EA%B7%80%EB%B6%84%EC%84%9D(Multiple_Regression_Analysis).ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Basic Pip and Install

In [ ]:
!pip install pingouin
!pip install --upgrade pingouin
!pip install scikit_posthocs

In [2]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [3]:
import pandas as pd 
import numpy as np 

import os
import sys
import warnings
warnings.filterwarnings('ignore')

import sklearn 
import missingno as msno
import matplotlib.pyplot as plt
%matplotlib inline
import seaborn as sns
plt.style.use('seaborn')
sns.set(font_scale=2.5)

import scipy 
from scipy import  stats
import statsmodels.api as sm
import pylab as py

import pingouin as pg
import scikit_posthocs

pd.options.display.float_format = '{:.3f}'.format

# 다중회귀분석(Multiple Regression Analysis)

단순회귀분석은 
- 단순 회귀 분석은 독립변수x 가 하나인 경우에 사용된다
다중회귀 분석은 
- 독립변수가 둘 이상인 경우에 사용된다. 

---



In [6]:
hos = sns.load_dataset('taxis')
hos

pickup              dropoff  passengers  distance   fare  \
0     2019-03-23 20:21:09  2019-03-23 20:27:24           1     1.600  7.000   
1     2019-03-04 16:11:55  2019-03-04 16:19:00           1     0.790  5.000   
2     2019-03-27 17:53:01  2019-03-27 18:00:25           1     1.370  7.500   
3     2019-03-10 01:23:59  2019-03-10 01:49:51           1     7.700 27.000   
4     2019-03-30 13:27:42  2019-03-30 13:37:14           3     2.160  9.000   
...                   ...                  ...         ...       ...    ...   
6428  2019-03-31 09:51:53  2019-03-31 09:55:27           1     0.750  4.500   
6429  2019-03-31 17:38:00  2019-03-31 18:34:23           1    18.740 58.000   
6430  2019-03-23 22:55:18  2019-03-23 23:14:25           1     4.140 16.000   
6431  2019-03-04 10:09:25  2019-03-04 10:14:29           1     1.120  6.000   
6432  2019-03-13 19:31:22  2019-03-13 19:48:02           1     3.850 15.000   

       tip  tolls  total   color      payment            pickup_zone  \
0    2.150  0.000 12.950  yellow  credit card        Lenox Hill West   
1    0.000  0.000  9.300  yellow         cash  Upper West Side South   
2    2.360  0.000 14.160  yellow  credit card          Alphabet City   
3    6.150  0.000 36.950  yellow  credit card              Hudson Sq   
4    1.100  0.000 13.400  yellow  credit card           Midtown East   
...    ...    ...    ...     ...          ...                    ...   
6428 1.060  0.000  6.360   green  credit card      East Harlem North   
6429 0.000  0.000 58.800   green  credit card                Jamaica   
6430 0.000  0.000 17.300   green         cash    Crown Heights North   
6431 0.000  0.000  6.800   green  credit card          East New York   
6432 3.360  0.000 20.160   green  credit card            Boerum Hill   

                          dropoff_zone pickup_borough dropoff_borough  
0                  UN/Turtle Bay South      Manhattan       Manhattan  
1                Upper West Side South      Manhattan       Manhattan  
2                         West Village      Manhattan       Manhattan  
3                       Yorkville West      Manhattan       Manhattan  
4                       Yorkville West      Manhattan       Manhattan  
...                                ...            ...             ...  
6428              Central Harlem North      Manhattan       Manhattan  
6429  East Concourse/Concourse Village         Queens           Bronx  
6430                    Bushwick North       Brooklyn        Brooklyn  
6431      East Flatbush/Remsen Village       Brooklyn        Brooklyn  
6432                   Windsor Terrace       Brooklyn        Brooklyn  

[6433 rows x 14 columns]

In [17]:
!pip install --upgrade patsy
!pip install patsy

In [20]:
import statsmodels.formula.api as smf
from patsy import dmatrices, dmatrix, demo_data
from statsmodels.stats.outliers_influence import variance_inflation_factor

model2 = smf.ols(formula='total~distance+fare+tip', data=hos).fit()

model2.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  total   R-squared:                       0.988
Model:                            OLS   Adj. R-squared:                  0.988
Method:                 Least Squares   F-statistic:                 1.728e+05
Date:                Wed, 20 Apr 2022   Prob (F-statistic):               0.00
Time:                        13:27:52   Log-Likelihood:                -11860.
No. Observations:                6433   AIC:                         2.373e+04
Df Residuals:                    6429   BIC:                         2.376e+04
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept      2.6478      0.032     83.140      0.000       2.585       2.710
distance       0.1762      0.013     13.852      0.000       0.151       0.201
fare           0.9811      0.004    227.718      0.000       0.973       0.990
tip            1.2597      0.009    141.097      0.000       1.242       1.277
==============================================================================
Omnibus:                     1609.002   Durbin-Watson:                   1.617
Prob(Omnibus):                  0.000   Jarque-Bera (JB):            27083.182
Skew:                           0.749   Prob(JB):                         0.00
Kurtosis:                      12.940   Cond. No.                         31.0
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

- 3파트가 출력 되었다. 모형검정, 계수검정, 잔차검정 
- 98% 의 Adj. R-squared 값으로 해석가능하고
- p값이 0 이므로 유의미하며 각각 total과  2.6, 0.17 , 0.98의 관계를 가진다. (정비례)
- Durbin-Watson:	1.617이므로 2에 가깝다. (잔차검정) - 잔차의 독립성 가정이 인정되므로 유의미하다 

Dep. Variable: 종속변수

R-squared: 결정계수 - **모형의 설명력** x100해서 측정 

Model: 분석모형

Adj. R-squared: 조정된 결정계수 **모형의 설명력** x100해서 측정 - 독립변수가 많아지면서 패널티가 주어지는 것 

Method: 모형 계산방식

F-statistics: 모형 적합도 통계량

Prob(F-statistics): F 통계량의 유의확률

DF Residuals: 잔차의 자유도

DF Model: 모형의 매개변수 수

AIC/BIC: 여러 모형의 적합도 비교

std err: 계수 추정치의 표준오차

t: t-value 검정통계량

P>|t|: 유의확률 - p value가 매우 작아야 모형이 적합하다 - 계수도 적합하다. 

[0.025 0.975]: 95% 신뢰구간 범위

Omnibus: 잔차의 왜도/첨도 검정통계량

Prob(Omnibus): Omnibus의 유의확률

Skew/Kurtosis: 왜도/첨도

Durbin-Watson: 잔차의 독립성 검정통계량

Jarque-Bera(JB): 잔차의 왜도/첨도 검정통계량

Prob(JB): JB의 유의확률

Cond. No.: 다중공선성 검정통계량

- 종속변수, 결정계수, 분석모형, 조정된 결정계수, 모형 계산방식,모형 적합도 통계량, F 통계량의 유의확률, 잔차의 자유도,모형의 매개변수 수, 여러 모형의 적합도 비교가 출력되고



- 계수 추정치의 표준오차, 검정통계량,유의확률, 95% 신뢰구간 범위가 출력되고



- 잔차의 왜도/첨도/ 검정통계량, Omnibus의 유의확률, 왜도/첨도, 잔차의 독립성 검정통계량, 잔차의 왜도/첨도 검정통계량, JB의 유의확률, 다중공선성 검정통계량이 출력된다.